katello-attach-subscription

```diff
@@ -99,70 +139,659 @@ if @yaml.has_key?(:settings) and @yaml[:settings].is_a?(Hash)
+              exit 1
+            end
+          end
+        else
+          puts "    No subscription found attached to #{system['name']}. Nothing to do."
+        end
+      end
+    else
+      if @options[:verbose]
+        puts "  VERBOSE: No empty hypervisor with valid subscription found."
+      end
+    end
+  end
+
+end
+
+
+# main function
+def vdcupdate()
+  # satellite url has to start with https or PUT will fail with http error
+  unless @options[:uri].start_with?('https://')
+    abort "FATAL ERROR: the uri must start with https://"
+  end
+  @default_type = nil
+  @supports_registered_by = false
   # initialize variables
   systems = []
+  guests_systems = []
   system_details = {}
+  hosts_data = []
+  hosts_details_data = {}
   subs = {}
   page = 0
   req = nil
   cachefile = @options[:cachefile].to_s + "_org" + @options[:org].to_s
-
   # Fill systems array from API of satellite. Check for cache usage.
-  if @options[:usecache] and File.file?(cachefile)
-    begin
-      cacheyaml = YAML.load_file(cachefile)
-      systems = cacheyaml['systems']
-      system_details = cacheyaml['system_details']
-    rescue Exception
-      abort " FATAL ERROR: Failed to read cache file. re-run without --use-cache."
-    end
+  if @options[:usecache]
+    puts "Starting host collection from cache. Please be patient."
+    systems = readfromcache('systems')
+    system_details = readfromcache('system_details')
   else
-    # no cache wanted or no cache file exists
-    # checking all of the systems 100 at the time, from page 0 to latest
-    while (page == 0 or req['results'].length == req['per_page'].to_i)
-      # increase page counter
-      page += 1
-      # get 100 results
-      # filter to be applied: organization, and search options
-      req = api.resource(@host_resource).call(:index, {:organization_id => @options[:org], :page => page, :per_page => 100, :search => @options[:search]})
-      # concatenate output - all of the results
-      systems.concat(req['results'])
+    # no cache wanted
+    puts "Starting host collection from API. Please be patient."
+    # preparing search_options to be add at parameters to api call
+    # if search options are present check string to become @options[:search] to 'and #{@options[:search]}'
+    # final search_options would become 'hypervisor=true' or 'hypervisor=true and @options[:search]'
+    if @options[:search].to_s.length > 0
+      @options[:search] = " and #{@options[:search]}"
     end
+    # checking all of the systems 100 at the time, from page 0 to latest
+    # first of all, search for hypervisor as we would like to subscribe in order hypervisors, physicals and guests
+    # Currently there is a BZ opened as is not possible to absolute search hypervisor, physical and guest (BZ1635861)
+    search_options = "hypervisor=true#{@options[:search]}"
+    systems = fetch_all_results(:hosts, :index, {:organization_id => @options[:org], :search => search_options})
```

Comment:
thanks for using `fetch_all_results`, it seems like we want to move the `organization_id` part also inside that function, what do you think?